controller/query.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/fabric8-services/fabric8-wit/errors"
+	"github.com/fabric8-services/fabric8-wit/jsonapi"
+	"github.com/fabric8-services/fabric8-wit/log"
+	"github.com/fabric8-services/fabric8-wit/login"
+	"github.com/fabric8-services/fabric8-wit/query"
+	"github.com/fabric8-services/fabric8-wit/rest"
+	"github.com/fabric8-services/fabric8-wit/space"
+	"github.com/goadesign/goa"
+)
+
+// QueryController implements the query resource.
+type QueryController struct {
+	*goa.Controller
+	db     application.DB
+	config QueryControllerConfiguration
+}
+
+// QueryControllerConfiguration the configuration for the QueryController
+type QueryControllerConfiguration interface {
+	GetCacheControlQueries() string
+	GetCacheControlQuery() string
+}
+
+// NewQueryController creates a query controller.
+func NewQueryController(service *goa.Service, db application.DB, config QueryControllerConfiguration) *QueryController {
+	return &QueryController{
+		Controller: service.NewController("QueryController"),
+		db:         db,
+		config:     config,
+	}
+}
+
+// Create runs the create action.
+func (c *QueryController) Create(ctx *app.CreateQueryContext) error {
+	currentUserIdentityID, err := login.ContextIdentity(ctx)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	return application.Transactional(c.db, func(appl application.Application) error {
+		err = appl.Spaces().CheckExists(ctx, ctx.SpaceID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		q := query.Query{
+			SpaceID: ctx.SpaceID,
+			Fields:  ctx.Payload.Data.Attributes.Fields,
+			Title:   strings.TrimSpace(ctx.Payload.Data.Attributes.Title),
+			Creator: *currentUserIdentityID,
+		}
+		dup, err := appl.Queries().IsDuplicate(ctx, &q)
+		if err != nil || dup {
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		err = appl.Queries().Create(ctx, &q)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		res := &app.QuerySingle{
+			Data: ConvertQuery(appl, ctx.Request, q),
+		}
+		ctx.ResponseData.Header().Set("Location", rest.AbsoluteURL(ctx.Request, app.QueryHref(ctx.SpaceID, res.Data.ID)))
+		return ctx.Created(res)
+	})
+}
+
+// ConvertQuery converts from internal to external REST representation
+func ConvertQuery(appl application.Application, request *http.Request, q query.Query) *app.Query {
+	queryType := query.APIStringTypeQuery
+	spaceID := q.SpaceID.String()
+	relatedURL := rest.AbsoluteURL(request, app.QueryHref(spaceID, q.ID))
+	creatorID := q.Creator.String()
+	userType := APIStringTypeUser
+	relatedCreatorLink := rest.AbsoluteURL(request, fmt.Sprintf("%s/%s", usersEndpoint, creatorID))
+	spaceRelatedURL := rest.AbsoluteURL(request, app.SpaceHref(spaceID))
+	appQuery := &app.Query{
+		Type: queryType,
+		ID:   &q.ID,
+		Attributes: &app.QueryAttributes{
+			Title:     q.Title,
+			Fields:    q.Fields,
+			CreatedAt: &q.CreatedAt,
+		},
+		Links: &app.GenericLinks{
+			Self:    &relatedURL,
```

Comment:
You might as well use the `ptr.String` function to get pointer values on-the-fly.

Reply:
I am using ptr package here https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/c234bcd1f0b1fe7131b360c25bd9707386f3ede5
but is it okay to use a variable if I am using it more than once?

Reply:
Sure, then I would use a variable too. +1